diff -ruN .gitignore .gitignore
--- .gitignore	1969-12-31 16:00:00.000000000 -0800
+++ .gitignore	2022-08-17 19:56:23.000000000 -0700
@@ -0,0 +1,2 @@
+*
+!.gitignore
diff -ruN compound/MorphoGovernance.sol compound/MorphoGovernance.sol
--- compound/MorphoGovernance.sol	2022-08-09 10:29:45.000000000 -0700
+++ compound/MorphoGovernance.sol	2022-08-17 19:56:23.000000000 -0700
@@ -121,7 +121,7 @@
         uint256 _maxSortedUsers,
         address _cEth,
         address _wEth
-    ) external initializer {
+    ) public initializer { // HARNESS
         __ReentrancyGuard_init();
         __Ownable_init();
 
diff -ruN compound/MorphoUtils.sol compound/MorphoUtils.sol
--- compound/MorphoUtils.sol	2022-08-09 10:29:45.000000000 -0700
+++ compound/MorphoUtils.sol	2022-08-17 19:59:55.000000000 -0700
@@ -4,7 +4,7 @@
 import "@rari-capital/solmate/src/utils/SafeTransferLib.sol";
 import "@openzeppelin/contracts/utils/math/Math.sol";
 import "./libraries/CompoundMath.sol";
-import "@morpho-labs/morpho-utils/DelegateCall.sol";
+import "../../helpers/DelegateCallHarness.sol";
 
 import "./MorphoStorage.sol";
 
diff -ruN compound/PositionsManager.sol compound/PositionsManager.sol
--- compound/PositionsManager.sol	2022-08-09 10:29:45.000000000 -0700
+++ compound/PositionsManager.sol	2022-08-17 19:56:23.000000000 -0700
@@ -212,7 +212,7 @@
         address _onBehalf,
         uint256 _amount,
         uint256 _maxGasForMatching
-    ) external {
+    ) external override { //HARNESS : get a compiler error without this
         if (_onBehalf == address(0)) revert AddressIsZero();
         if (_amount == 0) revert AmountIsZero();
         _updateP2PIndexes(_poolToken);
@@ -302,7 +302,7 @@
         address _poolToken,
         uint256 _amount,
         uint256 _maxGasForMatching
-    ) external {
+    ) external override { // HARNESS
         if (_amount == 0) revert AmountIsZero();
         _updateP2PIndexes(_poolToken);
 
@@ -400,7 +400,7 @@
         address _supplier,
         address _receiver,
         uint256 _maxGasForMatching
-    ) external {
+    ) external override { // HARNESS
         if (_amount == 0) revert AmountIsZero();
         if (!userMembership[_poolToken][_supplier]) revert UserNotMemberOfMarket();
 
@@ -424,7 +424,7 @@
         address _onBehalf,
         uint256 _amount,
         uint256 _maxGasForMatching
-    ) external {
+    ) external override { // HARNESS
         if (_amount == 0) revert AmountIsZero();
         if (!userMembership[_poolToken][_onBehalf]) revert UserNotMemberOfMarket();
 
@@ -444,7 +444,7 @@
         address _poolTokenCollateral,
         address _borrower,
         uint256 _amount
-    ) external {
+    ) external virtual override {
         if (
             !userMembership[_poolTokenBorrowed][_borrower] ||
             !userMembership[_poolTokenCollateral][_borrower]
