diff -ruN compound/MorphoGovernance.sol compound/MorphoGovernance.sol
--- compound/MorphoGovernance.sol	2022-05-11 13:56:34.469985672 +0200
+++ compound/MorphoGovernance.sol	2022-05-11 14:17:21.269977337 +0200
@@ -120,7 +120,7 @@
         uint256 _maxSortedUsers,
         address _cEth,
         address _wEth
-    ) external initializer {
+    ) public initializer {
         __ReentrancyGuard_init();
         __Ownable_init();
 
diff -ruN .gitignore .gitignore
--- .gitignore	1970-01-01 01:00:00.000000000 +0100
+++ .gitignore	2022-05-11 14:15:03.159978260 +0200
@@ -0,0 +1,2 @@
+*
+!.gitignore
