diff -ruN compound/MorphoGovernance.sol compound/MorphoGovernance.sol
--- compound/MorphoGovernance.sol	2022-09-27 19:01:55.593922693 +0200
+++ compound/MorphoGovernance.sol	2022-09-27 19:18:26.006755489 +0200
@@ -147,7 +147,7 @@
         uint256 _maxSortedUsers,
         address _cEth,
         address _wEth
-    ) external initializer {
+    ) public initializer { // HARNESS
         __ReentrancyGuard_init();
         __Ownable_init();
 
diff -ruN compound/MorphoUtils.sol compound/MorphoUtils.sol
--- compound/MorphoUtils.sol	2022-09-27 19:01:55.593922693 +0200
+++ compound/MorphoUtils.sol	2022-09-27 19:18:26.006755489 +0200
@@ -4,7 +4,7 @@
 import "@rari-capital/solmate/src/utils/SafeTransferLib.sol";
 import "@openzeppelin/contracts/utils/math/Math.sol";
 import "./libraries/CompoundMath.sol";
-import "@morpho-dao/morpho-utils/DelegateCall.sol";
+import "../../helpers/DelegateCallHarness.sol";
 
 import "./MorphoStorage.sol";
 
diff -ruN compound/PositionsManager.sol compound/PositionsManager.sol
--- compound/PositionsManager.sol	2022-09-27 19:01:55.593922693 +0200
+++ compound/PositionsManager.sol	2022-09-27 19:18:47.906718648 +0200
@@ -235,7 +235,8 @@
         address _onBehalf,
         uint256 _amount,
         uint256 _maxGasForMatching
-    ) external {
+    ) external override {
+        //HARNESS : get a compiler error without this
         if (_onBehalf == address(0)) revert AddressIsZero();
         if (_amount == 0) revert AmountIsZero();
         Types.MarketStatus memory market = marketStatus[_poolToken];
@@ -332,7 +333,8 @@
         address _poolToken,
         uint256 _amount,
         uint256 _maxGasForMatching
-    ) external {
+    ) external override {
+        // HARNESS
         if (_amount == 0) revert AmountIsZero();
         Types.MarketStatus memory market = marketStatus[_poolToken];
         if (!market.isCreated) revert MarketNotCreated();
@@ -435,7 +437,8 @@
         address _supplier,
         address _receiver,
         uint256 _maxGasForMatching
-    ) external {
+    ) external override {
+        // HARNESS
         if (_amount == 0) revert AmountIsZero();
         Types.MarketStatus memory market = marketStatus[_poolToken];
         if (!market.isCreated) revert MarketNotCreated();
@@ -462,7 +465,8 @@
         address _onBehalf,
         uint256 _amount,
         uint256 _maxGasForMatching
-    ) external {
+    ) external override {
+        // HARNESS
         if (_amount == 0) revert AmountIsZero();
         Types.MarketStatus memory market = marketStatus[_poolToken];
         if (!market.isCreated) revert MarketNotCreated();
@@ -485,7 +489,7 @@
         address _poolTokenCollateral,
         address _borrower,
         uint256 _amount
-    ) external {
+    ) external virtual override {
         Types.MarketStatus memory collateralMarket = marketStatus[_poolTokenCollateral];
         if (!collateralMarket.isCreated) revert MarketNotCreated();
         if (collateralMarket.isLiquidateCollateralPaused) revert LiquidateCollateralPaused();
diff -ruN .gitignore .gitignore
--- .gitignore	1970-01-01 01:00:00.000000000 +0100
+++ .gitignore	2022-09-27 19:18:26.006755489 +0200
@@ -0,0 +1,2 @@
+*
+!.gitignore
